Source/WebCore/Modules/url-pattern/URLPattern.cpp

```diff
@@ -26,29 +26,247 @@
+
+    if (!init.password.isNull())
+        result.password = canonicalizePassword(init.password, type);
+
+    if (!init.hostname.isNull()) {
+        auto hostResult = canonicalizeHost(init.hostname, type);
+
+        if (hostResult.hasException())
+            return hostResult.releaseException();
+
+        result.hostname = hostResult.releaseReturnValue();
+    }
+
+    if (!init.port.isNull()) {
+        auto portResult = canonicalizePort(init.port, init.protocol, type);
+
+        if (portResult.hasException())
+            return portResult.releaseException();
+
+        result.port = portResult.releaseReturnValue();
+    }
+
+    if (!init.pathname.isNull()) {
+        result.pathname = init.pathname;
+
+        if (!baseURL.isNull() && baseURL.hasOpaquePath() && !isAbsolutePathname(result.pathname, type)) {
+            String baseURLPath = processBaseURLString(baseURL.path(), type);
+            size_t slashIndex = baseURLPath.reverseFind('/');
+
+            if (slashIndex != notFound)
+                result.pathname = makeString(StringView { baseURLPath }.left(slashIndex + 1), result.pathname);
+
+            auto pathResult = canonicalizePath(result.pathname, baseURL.protocol(), type);
+
+            if (pathResult.hasException())
+                return pathResult.releaseException();
+
+            result.pathname = pathResult.releaseReturnValue();
+        }
+    }
+
+    if (!init.search.isNull()) {
+        auto queryResult = canonicalizeSearch(init.search, type);
+
+        if (queryResult.hasException())
+            return queryResult.releaseException();
+
+        result.search = queryResult.releaseReturnValue();
+    }
+
+    if (!init.hash.isNull()) {
+        auto fragmentResult = canonicalizeHash(init.hash, type);
+
+        if (fragmentResult.hasException())
+            return fragmentResult.releaseException();
+
+        result.hash = fragmentResult.releaseReturnValue();
+    }
+
+    return result;
+}
+
+ExceptionOr<Ref<URLPattern>> URLPattern::create(URLPatternInput&&, String&& baseURL, URLPatternOptions&&)
 {
     UNUSED_PARAM(baseURL);
 
+    // FIXME: Implement URLPattern::create
+
     return adoptRef(*new URLPattern());
 }
 
-Ref<URLPattern> URLPattern::create(std::optional<URLPatternInput>&&, URLPatternOptions&&)
+ExceptionOr<Ref<URLPattern>> URLPattern::create(std::optional<URLPatternInput>&& input, URLPatternOptions&&)
 {
-    return adoptRef(*new URLPattern());
+    URLPatternInit init;
+
+    if (std::holds_alternative<String>(input.value()) && !std::get<String>(input.value()).isNull()) {
+        // FIXME: Implement converting String input to URLPatternInit.
+    } else if (std::holds_alternative<URLPatternInit>(input.value()))
+        init = std::get<URLPatternInit>(input.value());
+
+    auto maybeProcessedInit = processInit(WTFMove(init), BaseURLStringType::Pattern, String(), String(), String(), String(), String(), String(), String(), String());
```

Comment:
Is passing `String()` like this for most parameters a temporary thing due to incomplete implementation or how it's going to be longer term? If that how it's going to be longer term, I would add default parameter values on the function so that the caller can simply omit those parameters. If it's temporary, I would add a FIXME comment.

Reply:
I'll add default parameters. I don't think the processInit is called anywhere but during the "create" function and "create" specifies that these fields should always be null.